ext/src/ruby_api/store.rs

```diff
@@ -95,6 +95,51 @@ impl Store {
     }
 }
 
+/// A wrapper around a Ruby Value that has a store context.
+/// Used in places where both Store or Caller can be used.
+#[derive(Debug)]
+pub enum StoreContextValue {
```

Comment:
IMO, it's nicer to match against Rust enums than strings for a couple of reasons:

1. You get exhaustiveness checks
2. Avoid a call into libruby
3. You avoid an unnecessary string allocation (although probably trivial, it's nice to know your performance is bounded by a integer compare operation)

The potential downsides:

1. You are duplicating state instead of deriving in a sense, but in this case i'd probably prefer Rust being the source of truth. It takes some questions off the table that I'd rather not think about... (i.e what happens if a user inherits from `Wasmtime::Module`?)
2. You may need an extra `u32` to represent the state ([actually, this can likely be avoided if we use `RTypedData` instead of `Value` to make use of the non-zeroness](https://github.com/matsadler/magnus/blob/main/src/r_typed_data.rs#L46))

Reply:
> against Rust enums than strings

Generally I'm questioning if the usage of an enum is what we want. I believe it would be nicer if we could avoid the introduction of this special, intermediate representation, and either: just use the `Value` representation or convert any special values (e.g. `Store`, `Caller`) into their Rust representations immediately after entering Rust and let the type system represent any invariants of what's possible with each (e.g. calling `context` `context_mut`), and just that consistently across modules/functions that require it.

Reply:
[And here is the solution I think I like](https://github.com/bytecodealliance/wasmtime-rb/commit/bdd8486d4f5439ff149366fc6f00e21592255338). It uses `RTypedData` instead of `Value` (to have a more specific constraint with `NonZeroValue`), and `TryFrom` implementation to avoid the enums everywhere. It still stores an enum, but it's always derived from the Ruby state.

Reply:
> It still stores an enum, but it's always derived from the Ruby state.

One of the thing that was bothering Saúl was that it'd be "easy" to pass in incorrect value when creating the enum. This proposal partially addresses that, I like it.

---

We also explored using 2 other approaches:
1. Using a trait. We couldn't get it to work because there are places where we don't have a concrete type and can't introduce a generic (e.g. the `init` function). Maybe that would've been possible with a Box, but that added allocations and runtime dispatch, so we dropped the idea.
2. Using a Struct to store a `Value` and a enum of `&Caller` or `&Store`. I _think_ that would've worked, but require introducing lifetimes in _many_ places for little gain in the end.

Reply:
> Using a Struct to store a Value and a enum of &Caller or &Store. I think that would've worked, but require introducing lifetimes in many places for little gain in the end.

That's essentially what `RTypedData` is. Only limitation is that it does not support generics (makes derivation much harder). Magnus also makes sure to call `rb_check_typeddata` when attempting to convert, so passing in incorrect data will error at runtime (which we need regardless of static type checks).